compiler/aarch64/codegen/OMRLinkage.cpp

```diff
@@ -157,46 +157,263 @@ OMR::ARM64::Linkage::numArgumentRegisters(TR_RegisterKinds kind)
 
-            TR::MemoryReference *stackMR = new (cg()->trHeapMemory()) TR::MemoryReference(stackPtr, parmCursor->getParameterOffset(), cg());
-            cursor = generateMemSrc1Instruction(cg(), op, NULL, stackMR, linkageReg, cursor);
+         // Copy to the ai register if necessary
+         //
+         if (isGlobalRegisterAssigned(paramCursor) && ai != sourceIndex)
+            {
+            // This parameter needs a RegReg move.  We don't know yet whether
+            // we need the value in the target register, so for now we just
+            // remember that we need to do this and keep going.
+            //
+            TR_ASSERT(movStatus[ai         ].sourceReg == noReg, "Each target reg must have only one source");
+            TR_ASSERT(movStatus[sourceIndex].targetReg == noReg, "Each source reg must have only one target");
+            if (debug("traceCopyParametersToHomeLocation"))
+               diagnostic("copyParametersToHomeLocation: Planning to move %d to %d\n", sourceIndex, ai);
+            movStatus[ai].sourceReg                  = sourceIndex;
+            movStatus[sourceIndex].targetReg         = ai;
+            movStatus[sourceIndex].type              = paramType;
+            }
+
+         if (debug("traceCopyParametersToHomeLocation") && ai == sourceIndex)
+            {
+            diagnostic("copyParametersToHomeLocation: Parameter #%d already in register %d\n", lri, ai);
             }
          }
       }
 
-   return cursor;
+   // Phase 2: Iterate through the parameters again to insert the RegReg moves.
+   //
+   for (paramCursor = paramIterator.getFirst();
+       paramCursor != NULL;
+       paramCursor = paramIterator.getNext())
+      {
+      if (!paramCursor->isParmPassedInRegister())
+         continue;
+
+      const int8_t lri = paramCursor->getLinkageRegisterIndex();
+      const TR::RealRegister::RegNum paramReg = paramCursor->getType().isFloatingPoint() ? 
+                                 properties.getFloatArgumentRegister(lri) : properties.getIntegerArgumentRegister(lri);
+
+      if (movStatus[paramReg].targetReg == noReg)
+         {
+         // This parameter does not need to be copied anywhere
+         if (debug("traceCopyParametersToHomeLocation"))
+            diagnostic("copyParametersToHomeLocation: Not moving %d\n", paramReg);
+         }
+      else
+         {
+         if (debug("traceCopyParametersToHomeLocation"))
+            diagnostic("copyParametersToHomeLocation: Preparing to move %d\n", paramReg);
+
+         // If a mov's target register is the source for another mov, we need
+         // to do that other mov first.  The idea is to find the end point of
+         // the chain of movs starting with paramReg and ending with a
+         // register whose current value is not needed; then do that chain of
+         // movs in reverse order.
+         //
+         TR_ASSERT(noReg == 0, "noReg must be zero (not %d) for zero-filled initialization to work", noReg);
```

Comment:
Redundant assertion.  You have `TR_ASSERT(noReg == 0, ...)` earlier in this function.

Reply:
Removed this assertion.